app/helpers/pdf_helper.rb

```diff
@@ -47,7 +50,65 @@ def output_file
       "tmp/",
       )
     PdfForms.new.fill_form(source_pdf_path, pdf_tempfile.path, hash_for_pdf)
-    pdf_tempfile
+
+    if respond_to? :nys_form_type
+      ny_pdf_with_cfa_barcode(pdf_tempfile)
+    else
+      pdf_tempfile
+    end
+  end
+
+  def ny_pdf_with_cfa_barcode(pdf_tempfile)
+    num_pages = PDF::Reader.new(pdf_tempfile.path).page_count
+    barcode_page_paths = (1..num_pages).map do |page_num|
+      barcode_string = nys_12digit_barcode_value(nys_form_type, page_num)
+      rect_params = barcode_overlay_rect
+      pdf = Prawn::Document.new
+      pdf.rectangle(*rect_params)  # draw a white rectangle over the default barcode
+      pdf.fill_color "ffffff"
+      pdf.fill
+      pdf.fill_color "000000"
+      pdf.text_box barcode_string, :at => [0, 26], :width => 117, :size => 8, :align => :center
+      barcode = generate_barcode(barcode_string)
+      barcode.to_pdf(pdf, width: 1.1, x: -20, y: 0, height: 18, bottom_margin: 0)
+      barcode_page_tempfile = Tempfile.new(
+        ["pg#{page_num}", ".pdf"],
+        "tmp/",
+        )
+      pdf.render_file(barcode_page_tempfile.path)
+      barcode_page_tempfile
+    end
+    pdftk_wrapper = PdfForms.new
+    barcode_pages_tempfile = Tempfile.new(
+      ["barcodes", ".pdf"],
+      "tmp/",
+      )
+    pdftk_wrapper.cat(*barcode_page_paths.map(&:path), barcode_pages_tempfile.path)
+
+    new_pdf_tempfile = Tempfile.new(
+      [source_pdf_name, "_with_replaced_barcodes.pdf"],
+      "tmp/",
+      )
+    pdftk_wrapper.multistamp(pdf_tempfile.path, barcode_pages_tempfile.path, new_pdf_tempfile.path)
+    new_pdf_tempfile
+  end
+
+  def nys_10digit_barcode_value(form_type, page_num)
+    vendor_source_code = "1963"
+    last_two_digits_of_tax_year = tax_year.to_s[-2..-1]
+    "#{form_type}#{page_num}#{last_two_digits_of_tax_year}#{vendor_source_code}"
+  end
+
+  def nys_12digit_barcode_value(form_type, page_num)
+    three_digit_page_num = "%03d" % page_num
+    vendor_source_code = "1963"
```

Comment:
Nit: this is also declared on line 97, consider moving to `Ny201VPdf` (or `MultiTenantService` if it has broader reusability)

Reply:
Moved to Rails config